plonky3/src/circuit_builder.rs

```diff
@@ -232,12 +240,34 @@ impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val>> Air<AB> for P
 
         // public constraints
         let pi_moved = pi.to_vec();
-        let local = matrix.row_slice(0);
+        let (local, next) = (matrix.row_slice(0), matrix.row_slice(1));
 
-        // constraining Pi * (Ci - pub[i]) = 0
         publics.iter().zip(pi_moved).enumerate().for_each(
-            |(index, ((_, col_id, _), public_value))| {
-                let selector = local[self.analyzed.commitment_count() + index];
+            |(index, ((_, col_id, row_id), public_value))| {
+                let (incrementor, inverse, selector) = (
+                    local[self.analyzed.commitment_count() + 3 * index],
```

Comment:
let's have named variables for these. Let's say you introduce `x` in the doc at the top of this file, then here you can introduce `x` and `x_prime`, etc